matter_server/server/device_controller.py

```diff
@@ -0,0 +1,391 @@
+"""Controller that Manages Matter devices."""
+
+from __future__ import annotations
+
+import asyncio
+from collections import deque
+from concurrent import futures
+from datetime import datetime
+from enum import IntEnum
+from functools import partial
+import logging
+import os
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Coroutine,
+    Deque,
+    Final,
+    Generator,
+    List,
+    Optional,
+    Tuple,
+    Type,
+    Union,
+    cast,
+)
+
+from chip.ChipDeviceCtrl import ChipDeviceController
+from chip.discovery import FilterType as DiscoveryFilterType
+from chip.exceptions import ChipStackError
+
+from matter_server.server.const import SCHEMA_VERSION
+
+from ..common.helpers.api import api_command
+from ..common.helpers.util import dataclass_from_dict, dataclass_to_dict
+from ..common.models.device import CommissionOption
+from ..common.models.error import (
+    NodeCommissionFailed,
+    NodeInterviewFailed,
+    NodeNotExists,
+)
+from ..common.models.events import EventType
+from ..common.models.message import (
+    CommandMessage,
+    ErrorResultMessage,
+    SuccessResultMessage,
+)
+from ..common.models.node import MatterNode
+
+if TYPE_CHECKING:
+    from chip.clusters import (
+        Attribute,
+        Cluster,
+        ClusterAttributeDescriptor,
+        ClusterEvent,
+        ClusterCommand
+    )
+
+    from .server import MatterServer
+    from .stack import MatterStack
+
+
+DATA_KEY_NODES = "nodes"
+DATA_KEY_LAST_NODEID = "last_nodeid"
+
+
+class MatterDeviceController:
+    """Class that manages the Matter devices."""
+
+    def __init__(
+        self,
+        server: MatterServer,
+    ):
+        self.server = server
+        self.logger = server.logger.getChild("device_controller")
+        # Instantiate the underlying ChipDeviceController instance on the Fabric
+        self.chip_controller: ChipDeviceController = (
+            server.stack.fabric_admin.NewController()
+        )
+        # we keep the last events in memory so we can include them in the diagnostics dump
+        self.event_history: Deque[Attribute.EventReadResult] = deque(maxlen=25)
+        self._subscriptions: dict[int, Attribute.SubscriptionTransaction] = {}
+        self._nodes: dict[int, MatterNode] = {}
+        self._wifi_creds_set = False
+        self.logger.debug("CHIP Device Controller Initialized")
```

Comment:
It's nice to not create side effects when creating instances. Maybe we don't need this log?

Reply:
Yeah, this log can be moved to the start function